File: hkust-robocon-mb/CAN/CAN转换器资料4421/A1+和A3动态链接库/动态库例程/DELPHI7/Unit1.pas
unit Unit1;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, StdCtrls;

type

  TForm1 = class(TForm)
    Button1: TButton;
    Button2: TButton;
    Button3: TButton;
    Button4: TButton;
    Button5: TButton;
    Memo1: TMemo;
    procedure Button1Click(Sender: TObject);
    procedure Button2Click(Sender: TObject);
    procedure Button3Click(Sender: TObject);
    procedure Button5Click(Sender: TObject);
    procedure Button4Click(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Form1: TForm1;
//以下是调用DLL声明
TYPE
    P_Init_can = function(com_NUM:BYTE;Model:BYTE;CanBaudRate:WORD;SET_ID_TYPE:BYTE;FILTER_MODE:BYTE;RXF:pointer;RXM:pointer):integer;stdcall;
    P_Quit_can = function():integer;stdcall;
    P_Can_send = function(IDbuff:pointer;Databuff:pointer;FreamType:BYTE;Bytes:BYTE):integer;stdcall;
    P_Can_receive = function(IDbuff:pointer;Databuff:pointer;FreamType:pointer;Bytes:pointer):integer;stdcall;
VAR
    LM_DLL:Thandle;
    Init_can : P_Init_can;
    Quit_can : P_Quit_can;
    Can_send : P_Can_send;
    Can_receive : P_Can_receive;
//以上是调用DLL声明

implementation

{$R *.dfm}

procedure TForm1.Button1Click(Sender: TObject);
begin
    //这里是动态加载 DLL函数例子。
    LM_DLL := LoadLibrary('LM_USB.dll'); //动态方式加载dll
    if (LM_DLL <> NULL) then
    begin
	      @Init_can := GetProcAddress(LM_DLL, 'Init_can');
        @Quit_can := GetProcAddress(LM_DLL, 'Quit_can');
        @Can_send := GetProcAddress(LM_DLL, 'Can_send');
        @Can_receive := GetProcAddress(LM_DLL, 'Can_receive');
        Memo1.Lines.Append('加载DLL成功');
    end
	  else
    begin
	       Memo1.Lines.Append('加载DLL失败');
    end;
    //这里是动态加载 DLL函数例子。
    Memo1.SelLength := Length(Memo1.Text);
end;

procedure TForm1.Button2Click(Sender: TObject);
var ret:integer;
    RXF,RXM:array [0..3] of BYTE;
begin
    //端口号,型号,CAN波特率,ID排列方法,屏蔽滤波状态,滤波寄存器,屏蔽寄存器
    ret:=Init_can(0,1,20,0,0,@RXF,@RXM); //初始化函数演示
    if (ret=0) then Memo1.Lines.Append('连接设备成功')
    else Memo1.Lines.Append('连接设备失败，返回值'+inttostr(ret));
    Memo1.SelLength := Length(Memo1.Text);
end;

procedure TForm1.Button3Click(Sender: TObject);
var
    IDbuff:array[0..3] of byte;
    Databuff:array [0..7] of byte;
    ret,i:integer;
    str:string;
begin
    IDbuff[0]:=0;
    IDbuff[1]:=0;
    Databuff[0]:=1;
    Databuff[1]:=2;
    Databuff[2]:=3;
    Databuff[3]:=4;
    Databuff[4]:=5;
    Databuff[5]:=6;
    Databuff[6]:=7;
    Databuff[7]:=8;
    ret:= Can_send(@IDbuff,@Databuff,0,8);//发送函数演示
    Memo1.Lines.Append('已发送 '+IntTostr(IDbuff[0])+'  '+IntTostr(IDbuff[1])+'    '+IntTostr(Databuff[0])+'  '+IntTostr(Databuff[1])+'  '+IntTostr(Databuff[2])+'  '+IntTostr(Databuff[3])+'  '+IntTostr(Databuff[4])+'  '+IntTostr(Databuff[5])+'  '+IntTostr(Databuff[6])+'  '+IntTostr(Databuff[7]));
    Memo1.SelLength := Length(Memo1.Text);
end;

procedure TForm1.Button5Click(Sender: TObject);
var ret:integer;
begin
    ret := Quit_can();  //释放函数演示。退出程序或需重设参数需设用此函数。
    Memo1.Lines.Append('释放设备连接');
    Memo1.SelLength := Length(Memo1.Text);
end;

procedure TForm1.Button4Click(Sender: TObject);
var ret:integer;
    no_data:integer;
    IDbuff:array[0..3] of byte;
    Databuff:array [0..7] of byte;
    FreamType:BYTE;
    Bytes:BYTE;
begin
  	no_data := 1;
		repeat
			  ret := Can_receive(@IDbuff,@Databuff,@FreamType,@Bytes);//调用接收函数例子
				if (ret=1) then
				begin
					 Memo1.Lines.Append('接收:'+IntTostr(IDbuff[0])+'  '+IntTostr(IDbuff[1])+'    '+IntTostr(Databuff[0])+'  '+IntTostr(Databuff[1])+'  '+IntTostr(Databuff[2])+'  '+IntTostr(Databuff[3])+'  '+IntTostr(Databuff[4])+'  '+IntTostr(Databuff[5])+'  '+IntTostr(Databuff[6])+'  '+IntTostr(Databuff[7]));
					 no_data := 0;
				end;
    until (ret<>1);
		if (no_data =1) then Memo1.Lines.Append('接收: 无数据');
    Memo1.SelLength := Length(Memo1.Text);
end;

end.

File: hkust-robocon-mb/CAN/CANF+¼µìósÖ¿F¦äµûÖ4421/A1+sÆîA3sè¿µÇüTô+µÄÑs¦ô/sè¿µÇüs¦ôS+ït¿ï/BCB60/Unit1.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop
   //以下是调用DLL声明
#include <windows.h>
#include <Sysdefs.h>
HINSTANCE LM_DLL;
typedef int (__stdcall *P_Init_can)(unsigned char com_NUM,unsigned char Model,unsigned int CanBaudRate,unsigned char SET_ID_TYPE,unsigned char FILTER_MODE,unsigned char RXF[],unsigned char RXM[]);
typedef	int	(__stdcall *P_Can_send)(unsigned char IDbuff[],unsigned char Databuff[],unsigned char FreamType,unsigned char Bytes);
typedef	int	(__stdcall *P_Can_receive)(unsigned char IDbuff[],unsigned char Databuff[],unsigned char *FreamType,unsigned char *Bytes);
typedef	int	(__stdcall *P_Quit_can)(void);
P_Init_can Init_can;
P_Can_send Can_send;
P_Can_receive Can_receive;
P_Quit_can Quit_can;
//以上是调用DLL声明


#include "Unit1.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
        : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TForm1::Button1Click(TObject *Sender)
{
    //这里是动态加载 DLL函数例子。
    LM_DLL = LoadLibrary("LM_USB.dll"); //动态方式加载dll

    if (LM_DLL != NULL)
    {
        Init_can = (P_Init_can)GetProcAddress(LM_DLL, "Init_can");
        Can_send = (P_Can_send)GetProcAddress(LM_DLL, "Can_send");
        Can_receive = (P_Can_receive)GetProcAddress(LM_DLL, "Can_receive");
        Quit_can = (P_Quit_can)GetProcAddress(LM_DLL, "Quit_can");
        Memo1->Lines->Append("加载DLL成功");
    }
    else Memo1->Lines->Append("加载DLL失败");
    //这里是动态加载 DLL函数例子。
    SendMessage(Memo1->Handle, WM_VSCROLL, MAKELONG(SB_BOTTOM, 0), 0);//显示最后一行
}
//---------------------------------------------------------------------------
void __fastcall TForm1::Button2Click(TObject *Sender)
{
    unsigned int ret;
    unsigned char RXF[4],RXM[4];
    RXF[0] = 0;
    RXF[1] = 0;
    RXF[2] = 0;
    RXF[3] = 0;
    RXM[0] = 0;
    RXM[1] = 0;
    RXM[2] = 0;
    RXM[3] = 0;
    ret = Init_can(0,1,20,0,0,RXF,RXM); //调用连接A1+设备函数例子
    //ret 返回值 0：正常  1：已连接设备  2：无应答（端口有效无应答）3：无可用的串口（3-30）
    if (ret==0) Memo1->Lines->Append("连接设备成功");
    else Memo1->Lines->Append("连接设备失败，返回值"+IntToStr(ret));
    SendMessage(Memo1->Handle, WM_VSCROLL, MAKELONG(SB_BOTTOM, 0), 0);//显示最后一行
}
//---------------------------------------------------------------------------
void __fastcall TForm1::Button3Click(TObject *Sender)
{
    unsigned int ret;
    unsigned char IDbuff[4],Databuff[8],FreamType,Bytes;
    IDbuff[0] = 0;
    IDbuff[1] = 0;
    Databuff[0] = 1;
    Databuff[1] = 2;
    Databuff[2] = 3;
    Databuff[3] = 4;
    Databuff[4] = 5;
    Databuff[5] = 6;
    Databuff[6] = 7;
    Databuff[7] = 8;
    FreamType = 0;//标准数据帧
    Bytes = 8; //本帧8个字节
    ret = Can_send(IDbuff,Databuff,FreamType,Bytes); //调用发送函数例子
    Memo1->Lines->Append("发送:"+IntToStr(IDbuff[0])+"  "+IntToStr(IDbuff[1])+"    "+IntToStr(Databuff[0])+"  "+IntToStr(Databuff[1])+"  "+IntToStr(Databuff[2])+"  "+IntToStr(Databuff[3])+"  "+IntToStr(Databuff[4])+"  "+IntToStr(Databuff[5])+"  "+IntToStr(Databuff[6])+"  "+IntToStr(Databuff[7]));
    SendMessage(Memo1->Handle, WM_VSCROLL, MAKELONG(SB_BOTTOM, 0), 0);//显示最后一行
}
//---------------------------------------------------------------------------
void __fastcall TForm1::Button4Click(TObject *Sender)
{
    unsigned int ret;
    unsigned char no_data;
    unsigned char IDbuff[4],Databuff[8],FreamType,Bytes;
    no_data = 1;
    do
    {
        ret = Can_receive(IDbuff,Databuff,&FreamType,&Bytes);//调用接收函数例子
	if (ret==1)
	{
	    Memo1->Lines->Append("接收:"+IntToStr(IDbuff[0])+"  "+IntToStr(IDbuff[1])+"    "+IntToStr(Databuff[0])+"  "+IntToStr(Databuff[1])+"  "+IntToStr(Databuff[2])+"  "+IntToStr(Databuff[3])+"  "+IntToStr(Databuff[4])+"  "+IntToStr(Databuff[5])+"  "+IntToStr(Databuff[6])+"  "+IntToStr(Databuff[7]));
	    no_data = 0;
	}
    } while (ret==1);
    if (no_data ==1) Memo1->Lines->Append("接收: 无数据");
    SendMessage(Memo1->Handle, WM_VSCROLL, MAKELONG(SB_BOTTOM, 0), 0);//显示最后一行
}
//---------------------------------------------------------------------------
void __fastcall TForm1::Button5Click(TObject *Sender)
{
    unsigned int ret;
    ret = Quit_can(); //调用释放函数例子
    Memo1->Lines->Append("释放设备连接");
    SendMessage(Memo1->Handle, WM_VSCROLL, MAKELONG(SB_BOTTOM, 0), 0);//显示最后一行
}


 //---------------------------------------------------------------------------

